### PR TITLE
Update: allow object methods in callback-return (fixes #4711)

### LINKED ISSUE #4711: callback-return does not work with object methods
I've read the documentation for the [rule](http://eslint.org/docs/rules/callback-return), but did not see anything related to this; nor could I find an issue.

**.eslintrc**
```json
{
	"env" : {
		"node" : true
	},

	"rules" : {
		"callback-return" : [ 2, [ "callback", "cb", "done", "next", "send.error", "send.success", "send.fail", "reply" ] ]
	}
}
```

**callback-return.js**

```js
'use strict'

( function ( item, send, callback ) {
	if ( item ) {
		send.fail();
	}

	send.success();

	if ( item ) {
		callback( item );
	}

	callback();
} );
```

**console**

```
~/Source/temp
> eslint -c .eslintrc callback-return.js

~/Source/temp/callback-return.js
  11:3  error  Expected return with your callback function  callback-return

✖ 1 problem (1 error, 0 warnings)
```

Given the options I supplied to the rule, I would expect ESLint to give an error for `send.fail();`.

Thoughts?

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/29149310-callback-return-does-not-work-with-object-methods?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by fakewaffle**

### CODE REVIEW

File: lib/rules/callback-return.js [hunk 1]
```diff
@@ -39,7 +39,15 @@ module.exports = function(context) {
      * @returns {Boolean} Whether or not this function matches our callback name.
      */
     function isCallback(node) {
-        return node.callee.type === "Identifier" && callbacks.indexOf(node.callee.name) > -1;
+        if (node.callee.type === "Identifier" && callbacks.indexOf(node.callee.name) > -1) {
+            return true;
+        }
+
+        if (node.callee.type === "MemberExpression" && callbacks.indexOf(node.callee.object.name + "." + node.callee.property.name) > -1) {
```
Comment (nzakas) [inline lib/rules/callback-return.js]: This assumes that both `object` and `property` are always identifiers, but they can be any number of types:

```js
foo.bar.baz;   // object is another MemberExpression
foo().baz;    // object is a CallExpression
```

So here we can either verify that both `object` and `property` are `Identifier` (and therefore, don't allow `foo.bar.baz` as a valid configuration), or expand this to take into account multiple nesting levels.

**Comment by nzakas** [comment]
Nice job on this, just some edge cases to consider.

File: lib/rules/callback-return.js (hunk 1 above)
Comment (fakewaffle) [inline lib/rules/callback-return.js]: Yeah, I just noticed that something like `reply( error.data ).code( 400 );` (from hapi) doesn't work either.

Comment (fakewaffle) [inline lib/rules/callback-return.js]: I'm leaning towards only allowing `foo.bar` and working towards adding the edge cases in a future pull request. I'll need them soon, but need this sooner.

Comment (mysticatea) [inline lib/rules/callback-return.js]: I recommend to use `sourceCode.getText(node.callee)`.
`new-cap` is using it for this.

Comment (nzakas) [inline lib/rules/callback-return.js]: `getText()` may not be safe because it includes comments, so that should be called out in the documentation

**Comment by nzakas** [comment]
@fakewaffle are you still working on this?

**Comment by fakewaffle** [comment]
Yeah! Just been busy of the holidays. :santa:

**Comment by nzakas** [comment]
No problem, do you know when you'll pick this up again? No pressure, just trying to plan the release.

**Comment by nzakas** [comment]
@fakewaffle just following up again. Do you have an ETA? (if not, we can have someone pick this up and finish it).

**Comment by fakewaffle** [comment]
Thanks for the follow up - and I'm sorry I didn't get back to you earlier.

What do you think is left to implement/address? [This](https://github.com/eslint/eslint/pull/4716/files#r47701267)?

**Comment by nzakas** [comment]
Yup

**Comment by ilyavolodin** [comment]
@fakewaffle Just poking again. Do you know when you might have time to finish this?

**Comment by ilyavolodin** [comment]
It's been over 3 month since the last update. Closing as it seems that this PR is not going anywhere:-( @fakewaffle Feel free to resubmit if you get around to finishing this work.

**Comment by josectello** [comment]
I'm looking through this PR to submit a modified PR with the more recent code, however could someone fill me in on what is still lacking? It would help a lot. :)

**Comment by BYK** [comment]
@josectello - I think it is just addressing the things in this discussion: https://github.com/eslint/eslint/pull/4716#discussion_r47701267

Btw. you can get the patch for this PR from https://patch-diff.githubusercontent.com/raw/eslint/eslint/pull/4716.diff and try applying that to latest master.

**Pull request closed without merging**